### ISSUE TITLE
NestedFolders: dashboards are not provisioned when foldersFromFilesStructure is enabled  

### ISSUE THREAD

**Issue by lorinescu** [grafana/grafana#73271]
### What happened?

Dashboards provisioning using the file provisioner with `foldersFromFilesStructure: true` and `nestedFolders` feature toggle enabled fails. The dashboards and folders do seem to exist in the database but the folder API endpoint returns 404/folder.notFound (ex: `http://localhost:3000/api/folders/dc71c6c1-cbc7-4b15-be16-266973022a67`)   . Turning off `foldersFromFilesStructure` provisions the dashboards at the top level with no folders.

### What did you expect to happen?

The folder structure and dashboards are created correctly like it happens when nestedFolders feature flag is turned off. 

### Did this work before?

It works only when nestedFolders feature toggle is disabled.

### How do we reproduce it?

```
# mkdir -p grafana_data/provisioning/dashboards/
# cat > grafana_data/provisioning/dashboards/config.yaml
apiVersion: 1
providers:
  - name: 'aaa'
    orgId: 1
    folder: ''
    folderUid: ''
    type: file
    disableDeletion: false
    updateIntervalSeconds: 10
    allowUiUpdates: false
    options:
      path: /var/lib/grafana/dashboards
      foldersFromFilesStructure: true

#  mkdir -p grafana_data/dashboards/some/folder
# cat > grafana_data/dashboards/some/folder/example.json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "type": "datasource",
        "uid": "grafana"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "cellHeight": "sm",
        "footer": {
          "countRows": false,
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "showHeader": true
      },
      "pluginVersion": "10.0.3",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "grafana"
          },
          "queryType": "list",
          "refId": "A"
        }
      ],
      "title": "Panel Title",
      "type": "table"
    }
  ],
  "refresh": "",
  "schemaVersion": 38,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "New dashboard",
  "version": 0,
  "weekStart": ""
}

# docker run --rm -p 3000:3000 -v ./grafana_data:/var/lib/grafana -e "GF_PATHS_PROVISIONING=/var/lib/grafana/provisioning"  --name=grafana grafana/grafana:10.0.3

In Grafana UI I see "some" folder with my dashboard

#  rm -rf grafana_data/{alerting,csv,grafana.db,plugins,png}
# docker run --rm -p 3000:3000 -v ./grafana_data:/var/lib/grafana -e "GF_PATHS_PROVISIONING=/var/lib/grafana/provisioning" -e "GF_FEATURE_TOGGLES_ENABLE=nestedFolders" --name=grafana grafana/grafana:10.0.3

In Grafana UI there are no folders or dashboards and the folder API endpoint returns 404/folder.notFound
```

### Environment (with versions)?

Grafana: 10.0.x
OS: Linux
Browser: Firefox, Safari, Orion

### Grafana platform?

Docker

**Comment by mellieA** [comment]
@lorinescu thanks for providing detailed steps. I'll ask our dev squad to take a look.

**Comment by welljonathan** [comment]
I came here to create an issue, but I believe this one is probably connected so I'll add some info here.

I'm using the standard helm chart with nested folders enabled. I'm trying to import that dashboards from the grafana backends (Loki in this example).  I set the the value of the folder attribute to `Teams/DevOps/Observability Stack/` and they show up correctly mounted in the sidecar. 

```
grafana-677668994b-8czhk:/tmp/dashboards$ ls -l Teams/DevOps/Observability\ Stack\ -\ Grafana\ Loki/
total 232
-rw-r--r--    1 grafana  472          21363 Aug 16 22:00 loki-chunks.json
-rw-r--r--    1 grafana  472           8564 Aug 16 22:00 loki-deletion.json
-rw-r--r--    1 grafana  472          14247 Aug 16 22:00 loki-logs.json
-rw-r--r--    1 grafana  472           9006 Aug 16 22:00 loki-mixin-recording-rules.json
-rw-r--r--    1 grafana  472          91868 Aug 16 22:00 loki-operational.json
-rw-r--r--    1 grafana  472          14725 Aug 16 22:00 loki-reads-resources.json
-rw-r--r--    1 grafana  472           9298 Aug 16 22:00 loki-reads.json
-rw-r--r--    1 grafana  472          22755 Aug 16 22:00 loki-retention.json
-rw-r--r--    1 grafana  472          10668 Aug 16 22:00 loki-writes-resources.json
-rw-r--r--    1 grafana  472           8265 Aug 16 22:00 loki-writes.json
grafana-677668994b-8czhk:/tmp/dashboards$

```

I see from the logs too that it found the foldersAnnotation value to put it in the right place in the mounted disk folder structure, but 

```
grafana-sc-dashboard {"time": "2023-08-16T23:01:32.280524+00:00", "msg": "Found a folder override annotation, placing the loki-dashboards-1 in: /tmp/dashboards/Teams/DevOps/Observability Stack - Grafana Loki", "level": "INFO"}
grafana-sc-dashboard {"time": "2023-08-16T23:01:32.283291+00:00", "msg": "Found a folder override annotation, placing the loki-dashboards-2 in: /tmp/dashboards/Teams/DevOps/Observability Stack - Grafana Loki", "level": "INFO"}
grafana-sc-dashboard {"time": "2023-08-16T23:02:37.305963+00:00", "msg": "Found a folder override annotation, placing the loki-dashboards-1 in: /tmp/dashboards/Teams/DevOps/Observability Stack - Grafana Loki", "level": "INFO"}
grafana-sc-dashboard {"time": "2023-08-16T23:02:37.308646+00:00", "msg": "Found a folder override annotation, placing the loki-dashboards-2 in: /tmp/dashboards/Teams/DevOps/Observability Stack - Grafana Loki", "level": "INFO"}
```

But, these files never actually make it into dashboards in the UI.

**Comment by leida94** [comment]
I also have the same issue. I use helm chart to publish Grafana to a kubernetes cluster and use sidecar to provision dashboard and dashboard folder tht way:
`sidecar:
  logLevel: info
  dashboards:
    enabled: true
    label: dashboard
    labelValue: generated
    folder: /var/lib/grafana/dashboards
    resource: configmap
    folderAnnotation: target-directory
    watchMethod: SLEEP

    provider:
      name: sidecarProvider
      foldersFromFilesStructure: true
      updateIntervalSeconds: 3600
`

When I enable the nested folder everything going to error. If I create a new provisioned dashboard folder (with dashboards) the Grafana cannot public the new item to the GUI until I restart the deployment (or the pod). (Naturally this function works properly without nested folder feature.)

I think the problem (or one of the problem) is the API call. When I would like to listing the dashboards **in the browser** use `
https://noting-interest-hostname/grafana/api/search?limit=1000`. But the side car and the provisioning use the dashboard folder api `GET /api/folders`.  If the nested folder feature enabled this two API calls results are not the same! 

Basically obvious this twist because if you do not open a folder (which contains nested folder(s)) in GUI the nested folder should be not listed the nested folder. But the folder API list all the folder nothing matter if it nested or not. 

Finally, my problem is same, if I would like to enable the nested folder feature (100% I want) the folder provisioning going to undefined behaviour.

**Comment by papagian** [comment]
related to #74671

**Comment by palmaarianne** [comment]
Looking forward to a fix for this one. We're having the same issues with our provisioned folders and dashboards as it seems the `foldersFromFilesStructure: true` doesn't work. Hope we can have some workaround as well. Thanks!

**Comment by QuentinBisson** [comment]
Is there a way to make this work? Because this prevents us from using nested folders

**Comment by papagian** [comment]
there is a draft that probably would fix it: #79793 but most probably needs to be updated since [Dashboards: Remove unique name constraints](https://github.com/grafana/grafana/pull/90687)

**Comment by jeremych1000** [comment]
We're on Grafana 8.6.4, and it seems to work?!?! The docs haven't been updated to say this yet.

Successfully provisioned dashboards via a configmap into subfolders. As above, there seems to be some weirdness with non unique folder names, e.g. it only considers the folder name, not the full folder path.

**Comment by jeremych1000** [comment]
> some weirdness with non unique folder names, e.g. it only considers the folder name, not the full folder path.

If I requested the folder as `foo/bar`, it creates it fine.
If I THEN request another folder as `baz/bar`, it ends up in `foo/bar`.

**Comment by liebsc21** [comment]
Are there any updates? We also require this feature.

**Comment by m4vr0x** [comment]
We still need this please do no close

**Comment by SlaughterBrute** [comment]
@undef1nd you're currently assigned to this issue. Is there any update?

**Comment by vfabi** [comment]
Any updates?
_Grafana v12.4.2 (ebade4c739)_
